enterprise/internal/campaigns/resolvers/changeset_connection_test.go

```diff
@@ -72,63 +79,157 @@ func TestChangesetConnectionResolver(t *testing.T) {
+		},
+		{
+			Typename: "HiddenExternalChangeset",
+			ID:       string(marshalChangesetID(changeset4.ID)),
+		},
+	}
 
-	input := map[string]interface{}{"campaign": campaignApiID}
-	var response struct{ Node apitest.Campaign }
-	apitest.MustExec(ctx, t, s, input, &response, queryChangesetConnection)
-
-	wantChangesets := apitest.ChangesetConnection{
-		Stats: apitest.ChangesetConnectionStats{
-			Unpublished: 1,
-			Open:        1,
-			Merged:      1,
-			Closed:      0,
-			Total:       3,
+	tests := []struct {
+		First           int
+		unsafeOpts      bool
+		wantHasNextPage bool
+		wantTotalCount  int
+		wantOpen        int
+		wantNodes       []apitest.Changeset
+	}{
+		{
+			First:           1,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:1],
+		},
+		{
+			First:           2,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:2],
+		},
+		{
+			First:           3,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:3],
+		},
+		{
+			First:           4,
+			wantHasNextPage: false,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:4],
+		},
+		{
+			First:           1,
+			unsafeOpts:      true,
+			wantHasNextPage: true,
+			// Expect only 3 changesets to be returned when an unsafe filter is applied.
+			wantTotalCount: 3,
+			wantOpen:       1,
+			wantNodes:      nodes[:1],
 		},
-		TotalCount: 3,
-		Nodes: []apitest.Changeset{
-			{
-				Typename:   "ExternalChangeset",
-				ID:         string(marshalChangesetID(changeset1.ID)),
-				Repository: apitest.Repository{Name: repo.Name},
-			},
-			{
-				Typename:   "ExternalChangeset",
-				ID:         string(marshalChangesetID(changeset2.ID)),
-				Repository: apitest.Repository{Name: repo.Name},
-			},
-			{
-				Typename:   "ExternalChangeset",
-				ID:         string(marshalChangesetID(changeset3.ID)),
-				Repository: apitest.Repository{Name: repo.Name},
-			},
+		{
+			First:           2,
+			unsafeOpts:      true,
+			wantHasNextPage: true,
+			wantTotalCount:  3,
+			wantOpen:        1,
+			wantNodes:       nodes[:2],
+		},
+		{
+			First:           3,
+			unsafeOpts:      true,
+			wantHasNextPage: false,
+			wantTotalCount:  3,
+			wantOpen:        1,
+			wantNodes:       nodes[:3],
 		},
 	}
 
-	if diff := cmp.Diff(wantChangesets, response.Node.Changesets); diff != "" {
-		t.Fatalf("wrong changesets response (-want +got):\n%s", diff)
+	reviewStatePending := campaigns.ChangesetReviewStatePending
+
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("Unsafe opts %t, first %d", tc.unsafeOpts, tc.First), func(t *testing.T) {
+			var reviewState *campaigns.ChangesetReviewState
+			if tc.unsafeOpts {
+				reviewState = &reviewStatePending
+			}
+			input := map[string]interface{}{"campaign": campaignApiID, "first": int64(tc.First), "reviewState": reviewState}
```

Comment:
Nitpick, but I'd do something like this:
```suggestion
			input := map[string]interface{}{"campaign": campaignApiID, "first": int64(tc.First)}
			if tc.unsafeOpts {
				input["reviewState"] = &reviewStatePending
			}
```